ASoC: SOF: Intel: Fix ROM/FW state reporting and handling

ujfalusi opened the pull request:
Hi,

it stared to bug me that in bug reports we seldom see the ROM error code decoded properly to human readable form we usually end up with the u32 hexa number...
It turned out that the code we have is not handling the FSR 'register' correctly and that is the reason why the missed decoding.
The code also checks for error codes in the FSR register, but that code is written to the _next_ register which is used by the ROM for error reporting and after firmware boot it is used by the firmware for status reporting.

This PR tries to fix this by introducing proper decoding of FSR's state, + wait_state, module and running areas and hopefully going to provide more usable reports for us, humans.

ujfalusi commented:
the checkpatch report for the first patch is bogus, imho.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -407,26 +404,141 @@ static const struct hda_dsp_msg_code hda_dsp_rom_msg[] = {
+	FSR_ROM_STATE_ENTRY(GET_LOAD_OFFSET),
+	FSR_ROM_STATE_ENTRY(FETCH_ROM_EXT),
+	FSR_ROM_STATE_ENTRY(FETCH_ROM_EXT_DONE),
+	/* CSE stares */
```
states?

ujfalusi replied:
haha, true.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -407,26 +404,141 @@ static const struct hda_dsp_msg_code hda_dsp_rom_msg[] = {
 	/* not for us, must be generic sof message */
-	dev_dbg(sdev->dev, "unknown ROM status value %8.8x\n", status);
+	if (!state_text) {
+		dev_dbg(sdev->dev, "0x%08x: unknown ROM status value\n", fsr);
```
%#08x?

ujfalusi replied:
should be `%#10x` or something, do you still prefer that form?

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -407,26 +404,141 @@ static const struct hda_dsp_msg_code hda_dsp_rom_msg[] = {
+			wait_state_text = "unknown";
+
+		dev_printk(level, sdev->dev,
+			   "0x%08x: module: %s, state: %s, waiting for: %s, %s\n",
```
%#

ujfalusi replied:
I want to have fixed size for the number, even if it is mostly 0 MSB, `%#10x` is the other option.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -407,26 +404,141 @@ static const struct hda_dsp_msg_code hda_dsp_rom_msg[] = {
+			   fsr, module_text, state_text, wait_state_text,
+			   fsr & FSR_RUNNING ? "running" : "not running");
+	} else {
+		dev_printk(level, sdev->dev, "0x%08x: module: %s, state: %s, %s\n",
```
&#

ujfalusi replied:
same here.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -407,26 +404,141 @@ static const struct hda_dsp_msg_code hda_dsp_rom_msg[] = {
-	u32 status;
+	const struct sof_intel_dsp_desc *chip = get_chip_info(sdev->pdata);
+	u32 fsr, state, wait_state, module, error_code;
+	const char *module_text, *error_text;
```
error_text might be used uninitialized line 533

ujfalusi replied:
True!

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -407,26 +404,141 @@ static const struct hda_dsp_msg_code hda_dsp_rom_msg[] = {
+	FSR_MODULE_NAME_ENTRY(ROM_BYP),
+	FSR_MODULE_NAME_ENTRY(BASE_FW),
+	FSR_MODULE_NAME_ENTRY(LP_BOOT),
+	FSR_MODULE_NAME_ENTRY(BRNGUP),
```
BRINGUP ?

ujfalusi replied:
It is defined like this for ROM and I kept it for easier 'git grep'

plbossart replied:
in the defines you have BRINGUP, so I was wondering if there was a disconnect.

ujfalusi replied:
It would not compile, I have:
```c
#define FSR_MOD_BRNGUP				0x4
```

Don't ask me why leaving out one `I` is a good idea, but it is what it is.

plbossart reviewed (changes requested):
couple of typos, nit-picks, one variable not initialized but looks good otherwise @ujfalusi

ujfalusi commented:
Changes since v1:
- use `%#010x` for printing hexa numbers instead of `0x%08x`
- add simple helper to get the text out from struct hda_dsp_msg_code arrays
  - No need to initialize the text pointers
- typo fix

ujfalusi commented:
Changes since v2:
- correct decoding of bit31 in FSR: 0 - running, 1 - halted.

ujfalusi commented:
Changes since v2:
- typo fixed in commit messages (double a)
- downgraded the prints to dev_dbg_ratelimited() since this issue should not happen on user's machine, only during development.

plbossart commented:
@ranj063 would you mind looking at this PR?
